sw/cheri/tests/hyperram_test.cc

```diff
@@ -0,0 +1,294 @@
+#include "../common/uart-utils.hh"
+
+#include <cheri.hh>
+#include <platform-uart.hh>
+#include <stdint.h>
+#include <ds/xoroshiro.h>
+
+using namespace CHERI;
+
+const int RandTestBlockSize = 256;
+const int HyperramSize = (1024 * 1024) / 4;
+
+#define CC_BOLD "1"
+#define CC_RED "31"
+#define CC_GREEN "32"
+#define CC_RESET "0"
+
+void set_console_mode(Capability<volatile OpenTitanUart<>> uart, const char* cc) {
+  write_str(uart, "\x1b[");
+  write_str(uart, cc);
+  write_str(uart, "m");
+}
+
+void write_test_result(Capability<volatile OpenTitanUart<>> uart, int failures) {
+  if (failures == 0) {
+    set_console_mode(uart, CC_GREEN);
+    write_str(uart, "PASS!\n");
+  } else {
+    set_console_mode(uart, CC_RED);
+    write_str(uart, "FAIL!\n");
+  }
+  set_console_mode(uart, CC_RESET);
+}
+
+int rand_data_test_block(Capability<volatile uint32_t> hyperram_area,
+    ds::xoroshiro::P64R32& prng, uint32_t start_hr_addr) {
+  uint32_t write_values[RandTestBlockSize];
+  uint32_t read_values[RandTestBlockSize];
+
+  for(int i = 0;i < RandTestBlockSize;++i) {
+    write_values[i] = prng();
+  }
+
+  for(int i = 0;i < RandTestBlockSize;++i) {
+    hyperram_area[i + start_hr_addr] = write_values[i];
+  }
+
+  for(int i = 0;i < RandTestBlockSize;++i) {
+    read_values[i] = hyperram_area[i + start_hr_addr];
+  }
+
+  int failures = 0;
+
+  for(int i = 0;i < RandTestBlockSize;++i) {
+    if (read_values[i] != write_values[i]) {
+      ++failures;
+    }
+  }
+
+  return failures;
+}
+
+int rand_data_test_full(Capability<volatile uint32_t> hyperram_area,
+  ds::xoroshiro::P64R32& prng) {
+
+  int failures = 0;
+  for (uint32_t addr = 0;addr < HyperramSize; addr += RandTestBlockSize) {
+    failures += rand_data_test_block(hyperram_area, prng, addr);
+  }
+
+  return failures;
+}
+
+int rand_data_addr_test(Capability<volatile uint32_t> hyperram_area,
+  ds::xoroshiro::P64R32& prng, int iterations) {
+
+  int failures = 0;
+
+  for (int i = 0;i < iterations; ++i) {
+    uint32_t rand_addr;
+    uint32_t rand_val;
+    uint32_t read_val;
+
+    rand_addr = prng() % HyperramSize;
+    rand_val = prng();
+
+    hyperram_area[rand_addr] = rand_val;
+    read_val = hyperram_area[rand_addr];
+
+    if (read_val != rand_val) {
+      failures += 1;
+    }
+  }
+
+  return failures;
+}
+
+int rand_cap_test(Capability<volatile uint32_t> hyperram_area,
+    Capability<Capability<volatile uint32_t>> hyperram_cap_area,
+    ds::xoroshiro::P64R32& prng, int iterations, Capability<volatile OpenTitanUart<>> uart) {
+
+  int failures = 0;
+
+  for (int i = 0;i < iterations; ++i) {
+    uint32_t rand_addr;
+    uint32_t rand_cap_addr;
+    uint32_t rand_val;
+    uint32_t read_val;
+
+    Capability<volatile uint32_t> write_cap;
+    Capability<volatile uint32_t> read_cap;
+
+    do {
+      rand_addr = prng() % HyperramSize;
+      rand_cap_addr = prng() % HyperramSize;
+    } while ((rand_addr / 2) == (rand_cap_addr / 2));
+
+    rand_cap_addr /= 2;
```

Comment:
The intention would be clearer if written this way?
```suggestion
    do {
      rand_index = prng() % HyperramSize;
      rand_cap_index = prng() % (HyperramSize / 2); // Pointer is double word in size.
    } while (rand_index / 2 == rand_cap_index);
```

I think if a variable have name `addr`, it's probably better that it's in byte-addressed, otherwise it can be quite confusing.